types/wlr_text_input.c

```diff
@@ -0,0 +1,285 @@
+#ifndef _POSIX_C_SOURCE
+#define _POSIX_C_SOURCE 200809L
+#endif
+#include <assert.h>
+#include <stdlib.h>
+#include <stddef.h>
+#include <wlr/types/wlr_text_input.h>
+#include <wlr/util/log.h>
+#include "text-input-unstable-v3-protocol.h"
+#include "util/signal.h"
+
+static const struct zwp_text_input_v3_interface text_input_impl;
+
+static struct wlr_text_input *text_input_from_resource(
+		struct wl_resource *resource) {
+	assert(wl_resource_instance_of(resource, &zwp_text_input_v3_interface,
+		&text_input_impl));
+	return wl_resource_get_user_data(resource);
+}
+
+void wlr_text_input_send_enter(struct wlr_text_input *text_input,
+		struct wlr_surface *wlr_surface) {
+	zwp_text_input_v3_send_enter(text_input->resource, wlr_surface->resource);
+	text_input->focused_surface = wlr_surface;
+}
+
+void wlr_text_input_send_leave(struct wlr_text_input *text_input,
+		struct wlr_surface *wlr_surface) {
+	zwp_text_input_v3_send_leave(text_input->resource, wlr_surface->resource);
+	text_input->focused_surface = NULL;
+}
+
+void wlr_text_input_send_preedit_string(struct wlr_text_input *text_input,
+		const char *text, uint32_t cursor_begin, uint32_t cursor_end) {
+	zwp_text_input_v3_send_preedit_string(text_input->resource, text,
+		cursor_begin, cursor_end);
+}
+
+void wlr_text_input_send_commit_string(struct wlr_text_input *text_input,
+		const char *text) {
+	zwp_text_input_v3_send_commit_string(text_input->resource, text);
+}
+
+void wlr_text_input_send_delete_surrounding_text(
+		struct wlr_text_input *text_input, uint32_t before_length,
+		uint32_t after_length) {
+	zwp_text_input_v3_send_delete_surrounding_text(text_input->resource,
+		before_length, after_length);
+}
+
+void wlr_text_input_send_done(struct wlr_text_input *text_input) {
+	zwp_text_input_v3_send_done(text_input->resource,
+		text_input->current_serial);
+}
+
+static void wlr_text_input_destroy(struct wlr_text_input *text_input) {
+	wlr_signal_emit_safe(&text_input->events.destroy, text_input);
+	// remove from manager::text_inputs
+	wl_list_remove(&text_input->link);
+	// no need to destroy the resource if client closed, prompting the server to do it
+	if (text_input->resource) {
+		wl_resource_destroy(text_input->resource);
```

Comment:
We should never do this. Instead, the client should always send a request to destroy the resource.